File: frontend/src/utils/invoicePdfFilename.js
/**
 * Nom de fichier PDF facture personnalisé (téléchargement / onglet).
 * Ex. : Facture_Juillet_2026_EM-2026-07-0005_M_VUILLE_Michel_155CHF.pdf
 */

const MONTHS_FR = [
  'Janvier',
  'Fevrier',
  'Mars',
  'Avril',
  'Mai',
  'Juin',
  'Juillet',
  'Aout',
  'Septembre',
  'Octobre',
  'Novembre',
  'Decembre',
];

/**
 * @param {unknown} value
 * @param {number} [maxLen]
 * @returns {string}
 */
export function slugifyInvoiceFilenamePart(value, maxLen = 48) {
  if (value == null) return '';
  const ascii = String(value)
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .replace(/[^a-zA-Z0-9]+/g, '_')
    .replace(/_+/g, '_')
    .replace(/^_|_$/g, '');
  if (!ascii) return '';
  return ascii.slice(0, maxLen);
}

/**
 * @param {unknown} amount
 * @returns {string} ex. 155CHF ou 155_50CHF
 */
export function formatInvoiceAmountForFilename(amount) {
  const n = Number(amount);
  if (!Number.isFinite(n)) return '0CHF';
  const rounded = Math.round(n * 100) / 100;
  if (Math.abs(rounded - Math.round(rounded)) < 1e-9) {
    return `${Math.round(rounded)}CHF`;
  }
  return `${rounded.toFixed(2).replace('.', '_')}CHF`;
}

/**
 * Libellé client / payeur pour le nom de fichier.
 * @param {object|null|undefined} invoice
 * @returns {string}
 */
export function resolveInvoiceFilenameClientLabel(invoice) {
  if (!invoice || typeof invoice !== 'object') return 'Client';

  const billingParty = invoice.billing_party?.display_name;
  if (billingParty && String(billingParty).trim()) {
    return String(billingParty).trim();
  }

  const companyName = invoice.billed_to_company?.name;
  if (companyName && String(companyName).trim()) {
    return String(companyName).trim();
  }

  const billTo = invoice.bill_to_client;
  if (billTo) {
    if (billTo.is_institution && billTo.institution_name) {
      return String(billTo.institution_name).trim();
    }
    const last = String(billTo.last_name || '').trim();
    const first = String(billTo.first_name || '').trim();
    if (last || first) return [last, first].filter(Boolean).join(' ');
    if (billTo.username) return String(billTo.username).trim();
  }

  const client = invoice.client;
  if (client) {
    if (client.patient_display_name) {
      return String(client.patient_display_name).trim();
    }
    if (client.is_institution && client.institution_name) {
      return String(client.institution_name).trim();
    }
    const last = String(client.last_name || '').trim();
    const first = String(client.first_name || '').trim();
    if (last || first) return [last, first].filter(Boolean).join(' ');
    if (client.display_name) return String(client.display_name).trim();
    if (client.username) return String(client.username).trim();
  }

  return 'Client';
}

/**
 * @param {object|null|undefined} invoice
 * @returns {string}
 */
export function buildInvoicePdfDownloadFilename(invoice) {
  const monthIdx = Number(invoice?.period_month);
  const year = Number(invoice?.period_year);
  const monthLabel =
    Number.isInteger(monthIdx) && monthIdx >= 1 && monthIdx <= 12
      ? MONTHS_FR[monthIdx - 1]
      : 'Periode';
  const yearLabel = Number.isFinite(year) && year > 0 ? String(year) : '';

  const numberRaw = String(invoice?.invoice_number || '').trim();
  /** Conserver les tirets du n° (ex. EM-2026-07-0005), retirer seulement les caractères dangereux. */
  const numberPart =
    numberRaw
      .normalize('NFD')
      .replace(/[\u0300-\u036f]/g, '')
      .replace(/[/\\?%*:|"<>]/g, '-')
      .replace(/\s+/g, '_')
      .replace(/_+/g, '_')
      .replace(/^_|_$/g, '')
      .slice(0, 64) ||
    (invoice?.id != null ? `ID_${invoice.id}` : 'SansNumero');

  const clientPart =
    slugifyInvoiceFilenamePart(resolveInvoiceFilenameClientLabel(invoice), 56) ||
    'Client';

  const amountPart = formatInvoiceAmountForFilename(invoice?.total_amount);

  const parts = ['Facture', monthLabel];
  if (yearLabel) parts.push(yearLabel);
  parts.push(numberPart, clientPart, amountPart);

  return `${parts.join('_')}.pdf`;
}

File: backend/shared/invoice_pdf_filename.py
"""Nom de fichier PDF facture personnalisé (téléchargement / Content-Disposition)."""

from __future__ import annotations

import re
import unicodedata
from typing import Any

_MONTHS_FR = (
    "Janvier",
    "Fevrier",
    "Mars",
    "Avril",
    "Mai",
    "Juin",
    "Juillet",
    "Aout",
    "Septembre",
    "Octobre",
    "Novembre",
    "Decembre",
)


def slugify_invoice_filename_part(value: Any, *, max_len: int = 48) -> str:
    """Normalise un libellé pour un nom de fichier sûr (ASCII, underscores)."""
    if value is None:
        return ""
    normalized = unicodedata.normalize("NFKD", str(value))
    ascii_text = normalized.encode("ascii", "ignore").decode("ascii")
    slug = re.sub(r"[^A-Za-z0-9]+", "_", ascii_text).strip("_")
    slug = re.sub(r"_+", "_", slug)
    return slug[:max_len] if slug else ""


def format_invoice_amount_for_filename(amount: Any) -> str:
    """Ex. 155CHF ou 155_50CHF."""
    try:
        value = float(amount)
    except (TypeError, ValueError):
        return "0CHF"
    rounded = round(value * 100) / 100
    if abs(rounded - round(rounded)) < 1e-9:
        return f"{round(rounded)}CHF"
    return f"{rounded:.2f}".replace(".", "_") + "CHF"


def _client_label_from_mapping(data: dict[str, Any] | None) -> str:
    if not data:
        return ""
    if data.get("is_institution") and data.get("institution_name"):
        return str(data["institution_name"]).strip()
    patient = (data.get("patient_display_name") or "").strip()
    if patient:
        return patient
    last = (data.get("last_name") or "").strip()
    first = (data.get("first_name") or "").strip()
    if last or first:
        return " ".join(p for p in (last, first) if p)
    for key in ("display_name", "username", "institution_name"):
        raw = (data.get(key) or "").strip()
        if raw:
            return raw
    return ""


def resolve_invoice_filename_client_label(invoice: Any) -> str:
    """Libellé client / payeur pour le nom de fichier."""
    billing_party = getattr(invoice, "billing_party", None)
    if billing_party is not None:
        name = (getattr(billing_party, "display_name", None) or "").strip()
        if name:
            return name

    billed_to = getattr(invoice, "billed_to_company", None)
    if billed_to is not None:
        name = (getattr(billed_to, "name", None) or "").strip()
        if name:
            return name

    bill_to = getattr(invoice, "bill_to_client", None)
    if bill_to is not None:
        is_inst = bool(getattr(bill_to, "is_institution", False))
        inst_name = (getattr(bill_to, "institution_name", None) or "").strip()
        if is_inst and inst_name:
            return inst_name
        user = getattr(bill_to, "user", None)
        last = (
            (getattr(user, "last_name", None) or "").strip() if user is not None else ""
        )
        first = (
            (getattr(user, "first_name", None) or "").strip()
            if user is not None
            else ""
        )
        if last or first:
            return " ".join(p for p in (last, first) if p)
        username = (
            (getattr(user, "username", None) or "").strip() if user is not None else ""
        )
        if username:
            return username

    # Repli via sérialisation légère si relations absentes
    try:
        serialized = invoice._serialize_client()
        label = _client_label_from_mapping(serialized)
        if label:
            return label
    except Exception:
        pass

    return "Client"


def build_invoice_pdf_download_filename(invoice: Any) -> str:
    """Ex. Facture_Juillet_2026_EM-2026-07-0005_VUILLE_Michel_155CHF.pdf"""
    try:
        month_idx = int(getattr(invoice, "period_month", 0) or 0)
    except (TypeError, ValueError):
        month_idx = 0
    try:
        year = int(getattr(invoice, "period_year", 0) or 0)
    except (TypeError, ValueError):
        year = 0

    month_label = _MONTHS_FR[month_idx - 1] if 1 <= month_idx <= 12 else "Periode"
    year_label = str(year) if year > 0 else ""

    number_raw = (getattr(invoice, "invoice_number", None) or "").strip()
    # Conserver les tirets du n° (ex. EM-2026-07-0005)
    if number_raw:
        normalized = unicodedata.normalize("NFKD", number_raw)
        ascii_num = normalized.encode("ascii", "ignore").decode("ascii")
        number_part = re.sub(r'[/\\?%*:|"<>]', "-", ascii_num)
        number_part = re.sub(r"\s+", "_", number_part).strip("_")[:64]
    else:
        number_part = ""
    if not number_part:
        invoice_id = getattr(invoice, "id", None)
        number_part = f"ID_{invoice_id}" if invoice_id is not None else "SansNumero"

    client_part = (
        slugify_invoice_filename_part(
            resolve_invoice_filename_client_label(invoice), max_len=56
        )
        or "Client"
    )
    amount_part = format_invoice_amount_for_filename(
        getattr(invoice, "total_amount", 0)
    )

    parts = ["Facture", month_label]
    if year_label:
        parts.append(year_label)
    parts.extend([number_part, client_part, amount_part])
    return f"{'_'.join(parts)}.pdf"
